### ISSUE TITLE
Source filtering in reindex API

### ISSUE THREAD

**Issue by utsavoza** [olivere/elastic#955]
I have been wanting to use source filtering in reindex API, but unfortunately I am unable to figure how it can be accomplished using elastic. I tried checking the both `reindex.go` and `reindex_test.go`, but couldn't determine if the source filtering is supported in the reindex API or not (I am trying to avoid using body map or string). What I am trying to achieve here is delineate the include and exclude fields in the source to filter out/in the fields during reindexing, something similar to:
```
{
  "source": {
    "index": "twitter",
    "_source": {
        "includes": [ "obj1.*", "obj2.*" ],
        "excludes": [ "*.description" ]
    },  
  },
  "dest": {
    "index": "new_twitter"
  }
}
```  

### Which version of Elastic are you using?

[x] elastic.v6 (for Elasticsearch 6.x)
[ ] elastic.v5 (for Elasticsearch 5.x)
[ ] elastic.v3 (for Elasticsearch 2.x)
[ ] elastic.v2 (for Elasticsearch 1.x)

**Comment by olivere** [comment]
Does the above query work, or is it documented somewhere? I didn't find anything like that with a quick glance over the Java source.

It seems source filtering does work on the query level, i.e. you specify a query that holds a [`FetchSourceContext`](https://github.com/olivere/elastic/blob/release-branch.v6/fetch_source_context.go) (named after the Java code). There are a few tests that illustrate its [standalone usage](https://github.com/olivere/elastic/blob/release-branch.v6/fetch_source_context_test.go) as well as its [usage on the query level](https://github.com/olivere/elastic/blob/d677dbe6d6a14f0b4f10dd28e906452f1e1d6d06/search_source_test.go#L84-L101).

**Comment by utsavoza** [comment]
I was referring to https://www.elastic.co/guide/en/elasticsearch/reference/current/docs-reindex.html which does mention `The source section supports all the elements that are supported in a search request.`. I had tried using the standalone usage but that still gets registered under `query` in the `source` object instead of `_source`.

**Comment by olivere** [comment]
Right, seems like the implementation is not correct.

Details for later:

`ReindexSource` needs to be changed to (basically) implement the `Source` part of [`ReindexRequestBuilder`](https://github.com/elastic/elasticsearch/blob/6.x/server/src/main/java/org/elasticsearch/index/reindex/ReindexRequestBuilder.java), which uses a [`SearchRequestBuilder`](https://github.com/elastic/elasticsearch/blob/6.x/server/src/main/java/org/elasticsearch/action/search/SearchRequestBuilder.java).

We have [`SearchSource`](https://github.com/olivere/elastic/blob/release-branch.v6/search_source.go), which is not quite the same. We better check [`SearchRequest`](https://github.com/olivere/elastic/blob/release-branch.v6/search_request.go) and compare it with the Java implementation, then expand on that and use it in `ReindexSource`.

Long story short, can you try to use a `Query` with `FetchSourceContext`, maybe that'll work as a workaround for now for your use case.

**Comment by olivere** [comment]
Will be part of 6.2.13, to be released soon.
